File: examples/test_container.cpp
#include "container_cpp/container.hpp"
#include <iostream>
#include <iomanip>
#include <cassert>

using namespace container;

void print_chunk(const Chunk& chunk, int index) {
    std::cout << "  Chunk " << index << ":\n";
    std::cout << "    Type: ";
    for (auto c : chunk.chunk_type) {
        std::cout << static_cast<char>(c);
    }
    std::cout << "\n";
    std::cout << "    Length: " << chunk.length << "\n";
    std::cout << "    CRC: 0x" << std::hex << std::setw(8) << std::setfill('0')
              << chunk.crc << std::dec << "\n";
}

int main() {
    // Create a container with 3 chunks
    Container original;
    original.num_chunks = 3;
    original.chunks.resize(3);

    // Chunk 0: IHDR
    original.chunks[0].length = 5;
    original.chunks[0].chunk_type = {'I', 'H', 'D', 'R'};
    original.chunks[0].data = {0x00, 0x00, 0x01, 0x00, 0x08};
    original.chunks[0].crc = 0x12345678;

    // Chunk 1: DATA
    original.chunks[1].length = 3;
    original.chunks[1].chunk_type = {'D', 'A', 'T', 'A'};
    original.chunks[1].data = {0xAA, 0xBB, 0xCC};
    original.chunks[1].crc = 0x87654321;

    // Chunk 2: IEND
    original.chunks[2].length = 0;
    original.chunks[2].chunk_type = {'I', 'E', 'N', 'D'};
    original.chunks[2].data = {};
    original.chunks[2].crc = 0xFFFFFFFF;

    std::cout << "Original container:\n";
    std::cout << "  Num chunks: " << original.num_chunks << "\n";
    for (size_t i = 0; i < original.chunks.size(); ++i) {
        print_chunk(original.chunks[i], i);
    }

    // Serialize
    Writer writer;
    original.write(writer);
    auto serialized = writer.finish();

    std::cout << "\nSerialized to " << serialized.size() << " bytes\n";

    // Expected: 4 + (4+4+5+4) + (4+4+3+4) + (4+4+0+4) = 48 bytes
    assert(serialized.size() == 48);

    // Deserialize
    Reader reader(serialized);
    auto parsed = Container::read(reader);

    std::cout << "\nParsed container:\n";
    std::cout << "  Num chunks: " << parsed.num_chunks << "\n";

    // Verify
    assert(parsed.num_chunks == original.num_chunks);
    assert(parsed.chunks.size() == original.chunks.size());

    for (size_t i = 0; i < original.chunks.size(); ++i) {
        print_chunk(parsed.chunks[i], i);

        assert(parsed.chunks[i].length == original.chunks[i].length);
        assert(parsed.chunks[i].chunk_type == original.chunks[i].chunk_type);
        assert(parsed.chunks[i].data == original.chunks[i].data);
        assert(parsed.chunks[i].crc == original.chunks[i].crc);
    }

    std::cout << "\nALL TESTS PASSED! C++ struct arrays work correctly.\n";
    return 0;
}

File: examples/test_container.py
#!/usr/bin/env python3
"""Test the container format with array of structs"""

import sys
sys.path.insert(0, 'examples/container_py')

from container import Chunk, Container

def test_container():
    # Create a container with 3 chunks
    chunks = [
        Chunk(
            length=5,
            chunk_type=[ord('I'), ord('H'), ord('D'), ord('R')],
            data=[0x00, 0x00, 0x01, 0x00, 0x08],
            crc=0x12345678
        ),
        Chunk(
            length=3,
            chunk_type=[ord('D'), ord('A'), ord('T'), ord('A')],
            data=[0xAA, 0xBB, 0xCC],
            crc=0x87654321
        ),
        Chunk(
            length=0,
            chunk_type=[ord('I'), ord('E'), ord('N'), ord('D')],
            data=[],
            crc=0xFFFFFFFF
        ),
    ]

    container = Container(
        num_chunks=3,
        chunks=chunks
    )

    print("Original container:")
    print(f"  Num chunks: {container.num_chunks}")
    for i, chunk in enumerate(container.chunks):
        print(f"  Chunk {i}:")
        print(f"    Type: {''.join(chr(c) for c in chunk.chunk_type)}")
        print(f"    Length: {chunk.length}")
        print(f"    CRC: 0x{chunk.crc:08x}")

    # Serialize
    serialized = container.write()
    print(f"\nSerialized to {len(serialized)} bytes")

    # Expected size:
    # 4 (num_chunks) + 3 * (4 + 4 + data_len + 4)
    # = 4 + 3 * 12 + (5 + 3 + 0) = 4 + 36 + 8 = 48 bytes
    expected_size = 4 + (4 + 4 + 5 + 4) + (4 + 4 + 3 + 4) + (4 + 4 + 0 + 4)
    assert len(serialized) == expected_size, f"Expected {expected_size} bytes, got {len(serialized)}"

    # Deserialize
    parsed, bytes_read = Container.read(serialized)
    print(f"\nParsed container ({bytes_read} bytes read):")
    print(f"  Num chunks: {parsed.num_chunks}")

    # Verify
    assert parsed.num_chunks == container.num_chunks, "Num chunks mismatch"
    assert len(parsed.chunks) == len(container.chunks), "Chunk count mismatch"

    for i in range(len(container.chunks)):
        orig = container.chunks[i]
        pars = parsed.chunks[i]
        print(f"  Chunk {i}:")
        print(f"    Type: {''.join(chr(c) for c in pars.chunk_type)}")
        print(f"    Length: {pars.length}")
        print(f"    CRC: 0x{pars.crc:08x}")

        assert pars.length == orig.length, f"Chunk {i} length mismatch"
        assert pars.chunk_type == orig.chunk_type, f"Chunk {i} type mismatch"
        assert pars.data == orig.data, f"Chunk {i} data mismatch"
        assert pars.crc == orig.crc, f"Chunk {i} CRC mismatch"

    print("\nALL TESTS PASSED! Struct arrays work correctly.")

if __name__ == "__main__":
    test_container()
